File: serial_link/framing/fcs.py
"""RFC 1662 FCS-16 (HDLC/PPP frame check sequence).

The 16-bit FCS uses the CRC-CCITT polynomial x^16 + x^12 + x^5 + 1 in its
*reflected* form (0x8408), exactly as specified in RFC 1662 Appendix C. The
256-entry lookup table there is reproduced here by generation rather than by
transcription, which removes any chance of a copy error.

Wire conventions (RFC 1662):
  * the running FCS is seeded with 0xFFFF;
  * the value transmitted is the *ones-complement* of the running FCS, sent
    low byte first (see :func:`frame_fcs`);
  * a receiver that runs :func:`fcs16` over ``info + fcs`` gets the constant
    "good FCS" residual 0xF0B8 when the frame is intact (see :func:`check_fcs`).
"""

INIT_FCS16 = 0xFFFF
GOOD_FCS16 = 0xF0B8


def _build_table():
    table = []
    for byte in range(256):
        crc = byte
        for _ in range(8):
            crc = (crc >> 1) ^ 0x8408 if crc & 1 else crc >> 1
        table.append(crc & 0xFFFF)
    return tuple(table)


FCSTAB = _build_table()


def fcs16(data, fcs=INIT_FCS16):
    """Return the running FCS-16 over ``data`` (not complemented).

    ``fcs`` may be passed to continue a computation across chunks.
    """
    for byte in data:
        fcs = (fcs >> 8) ^ FCSTAB[(fcs ^ byte) & 0xFF]
    return fcs


def frame_fcs(data):
    """Return the 2 FCS bytes to append after ``data``, low byte first."""
    fcs = fcs16(data) ^ 0xFFFF
    return bytes((fcs & 0xFF, (fcs >> 8) & 0xFF))


def check_fcs(data_with_fcs):
    """True if ``data_with_fcs`` (payload followed by its 2 FCS bytes) is intact."""
    return fcs16(data_with_fcs) == GOOD_FCS16

File: src/fcs.cpp
#include "serial_link/fcs.hpp"

namespace serial_link::fcs {

const std::array<std::uint16_t, 256>& table() {
    // Generated once from the reflected polynomial 0x8408 (RFC 1662 App. C).
    static const std::array<std::uint16_t, 256> tab = [] {
        std::array<std::uint16_t, 256> t{};
        for (int byte = 0; byte < 256; ++byte) {
            std::uint16_t crc = static_cast<std::uint16_t>(byte);
            for (int i = 0; i < 8; ++i) {
                crc = (crc & 1) ? static_cast<std::uint16_t>((crc >> 1) ^ 0x8408)
                                : static_cast<std::uint16_t>(crc >> 1);
            }
            t[static_cast<std::size_t>(byte)] = crc;
        }
        return t;
    }();
    return tab;
}

std::uint16_t fcs16(const std::uint8_t* data, std::size_t n, std::uint16_t fcs) {
    const auto& tab = table();
    for (std::size_t i = 0; i < n; ++i) {
        fcs = static_cast<std::uint16_t>((fcs >> 8) ^ tab[(fcs ^ data[i]) & 0xFF]);
    }
    return fcs;
}

std::uint16_t fcs16(const Bytes& data, std::uint16_t fcs) {
    return fcs16(data.data(), data.size(), fcs);
}

Bytes frame_fcs(const Bytes& data) {
    std::uint16_t fcs = static_cast<std::uint16_t>(fcs16(data) ^ 0xFFFF);
    return Bytes{static_cast<std::uint8_t>(fcs & 0xFF),
                 static_cast<std::uint8_t>((fcs >> 8) & 0xFF)};
}

bool check_fcs(const Bytes& data_with_fcs) {
    return fcs16(data_with_fcs) == GOOD_FCS16;
}

}  // namespace serial_link::fcs
